File: V2/client/response.go
package client

// PlayerResponse is the top-level response from the player API
type PlayerResponse struct {
	PlayabilityStatus PlayabilityStatus `json:"playabilityStatus"`
	VideoDetails      VideoDetails      `json:"videoDetails"`
	StreamingData     StreamingData     `json:"streamingData"`
}

// PlayabilityStatus indicates if the video can be played
type PlayabilityStatus struct {
	Status          string `json:"status"`
	Reason          string `json:"reason"`
	PlayableInEmbed bool   `json:"playableInEmbed"`
	LiveStreamability *LiveStreamability `json:"liveStreamability"`
}

// LiveStreamability contains live stream specific info
type LiveStreamability struct {
	LiveStreamabilityRenderer LiveStreamabilityRenderer `json:"liveStreamabilityRenderer"`
}

// LiveStreamabilityRenderer contains live stream renderer data
type LiveStreamabilityRenderer struct {
	VideoID      string `json:"videoId"`
	OfflineSlate struct {
		LiveStreamOfflineSlateRenderer struct {
			ScheduledStartTime string `json:"scheduledStartTime"`
		} `json:"liveStreamOfflineSlateRenderer"`
	} `json:"offlineSlate"`
}

// VideoDetails contains video metadata
type VideoDetails struct {
	VideoID          string             `json:"videoId"`
	Title            string             `json:"title"`
	LengthSeconds    string             `json:"lengthSeconds"`
	Keywords         []string           `json:"keywords"`
	ChannelID        string             `json:"channelId"`
	ShortDescription string             `json:"shortDescription"`
	Thumbnail        ThumbnailContainer `json:"thumbnail"`
	ViewCount        string             `json:"viewCount"`
	Author           string             `json:"author"`
	IsLiveContent    bool               `json:"isLiveContent"`
	IsPrivate        bool               `json:"isPrivate"`
	IsOwnerViewing   bool               `json:"isOwnerViewing"`
}

// ThumbnailContainer holds thumbnail data
type ThumbnailContainer struct {
	Thumbnails []ThumbnailData `json:"thumbnails"`
}

// ThumbnailData represents a single thumbnail
type ThumbnailData struct {
	URL    string `json:"url"`
	Width  int    `json:"width"`
	Height int    `json:"height"`
}

// StreamingData contains streaming format information
type StreamingData struct {
	ExpiresInSeconds string            `json:"expiresInSeconds"`
	Formats          []StreamingFormat `json:"formats"`
	AdaptiveFormats  []StreamingFormat `json:"adaptiveFormats"`
	HLSManifestURL   string            `json:"hlsManifestUrl"`
	DashManifestURL  string            `json:"dashManifestUrl"`
}

// StreamingFormat represents a playback format
type StreamingFormat struct {
	ITag             int        `json:"itag"`
	MimeType         string     `json:"mimeType"`
	URL              string     `json:"url"`
	SignatureCipher  string     `json:"signatureCipher"`

	Bitrate          int        `json:"bitrate"`
	AverageBitrate   int        `json:"averageBitrate"`
	ContentLength    string     `json:"contentLength"`

	Width            int        `json:"width"`
	Height           int        `json:"height"`
	FPS              int        `json:"fps"`
	Quality          string     `json:"quality"`
	QualityLabel     string     `json:"qualityLabel"`

	AudioQuality     string     `json:"audioQuality"`
	AudioChannels    int        `json:"audioChannels"`
	AudioSampleRate  string     `json:"audioSampleRate"`

	ApproxDurationMs string     `json:"approxDurationMs"`
	LastModified     string     `json:"lastModified"`
	ProjectionType   string     `json:"projectionType"`

	InitRange        *RangeData `json:"initRange"`
	IndexRange       *RangeData `json:"indexRange"`
}

// RangeData represents byte ranges for streaming
type RangeData struct {
	Start string `json:"start"`
	End   string `json:"end"`
}

File: TS/src/client/response.ts
// Response types for player API

// PlayerResponse is the top-level response from the player API
export interface PlayerResponse {
    playabilityStatus: PlayabilityStatus;
    videoDetails: VideoDetails;
    streamingData: StreamingData;
}

// PlayabilityStatus indicates if the video can be played
export interface PlayabilityStatus {
    status: string;
    reason?: string;
    playableInEmbed?: boolean;
    liveStreamability?: LiveStreamability;
}

// LiveStreamability contains live stream specific info
export interface LiveStreamability {
    liveStreamabilityRenderer: LiveStreamabilityRenderer;
}

// LiveStreamabilityRenderer contains live stream renderer data
export interface LiveStreamabilityRenderer {
    videoId: string;
    offlineSlate?: {
        liveStreamOfflineSlateRenderer: {
            scheduledStartTime?: string;
        };
    };
}

// VideoDetails contains video metadata
export interface VideoDetails {
    videoId: string;
    title: string;
    lengthSeconds: string;
    keywords?: string[];
    channelId: string;
    shortDescription: string;
    thumbnail: ThumbnailContainer;
    viewCount: string;
    author: string;
    isLiveContent: boolean;
    isPrivate: boolean;
    isOwnerViewing?: boolean;
}

// ThumbnailContainer holds thumbnail data
export interface ThumbnailContainer {
    thumbnails: ThumbnailData[];
}

// ThumbnailData represents a single thumbnail
export interface ThumbnailData {
    url: string;
    width: number;
    height: number;
}

// StreamingData contains streaming format information
export interface StreamingData {
    expiresInSeconds: string;
    formats?: StreamingFormat[];
    adaptiveFormats?: StreamingFormat[];
    hlsManifestUrl?: string;
    dashManifestUrl?: string;
}

// StreamingFormat represents a playback format
export interface StreamingFormat {
    itag: number;
    mimeType: string;
    url?: string;
    signatureCipher?: string;

    bitrate: number;
    averageBitrate?: number;
    contentLength?: string;

    width?: number;
    height?: number;
    fps?: number;
    quality: string;
    qualityLabel?: string;

    audioQuality?: string;
    audioChannels?: number;
    audioSampleRate?: string;

    approxDurationMs?: string;
    lastModified?: string;
    projectionType?: string;

    initRange?: RangeData;
    indexRange?: RangeData;
}

// RangeData represents byte ranges for streaming
export interface RangeData {
    start: string;
    end: string;
}
